torch/csrc/jit/fuser/compiler.cpp

```diff
@@ -144,6 +156,111 @@ static void setInputBroadcastGroups(KernelSpec& spec) {
+      "aten::neg(Tensor self) -> Tensor",
+      "aten::add(Tensor self, Tensor other, *, Scalar alpha) -> Tensor",
+      // add this used to be prim::AutogradAdd
+  }};
+  // Scan the graph. We will delete nodes. We want later (in the graph)
+  // _grad_sum_to_size nodes to have priority over earlier ones. Thus
+  // we scan backwards.
+  for (auto it = graph->nodes().rbegin(); it != graph->nodes().rend(); it++) {
+    auto* node = *it;
+    if (node->kind() == aten::_grad_sum_to_size) {
+      use_list uses_to_process(node->output()->uses());
+      while (!uses_to_process.empty()) {
+        auto user = uses_to_process[0].user;
+        auto offset = uses_to_process[0].offset;
+        uses_to_process.erase(uses_to_process.begin());
+        if (user->matches(
+                "aten::type_as(Tensor self, Tensor other) -> Tensor")) {
+          // sometimes, a mask or similar is cast to the same
+          // type as the gradient. But we only want to allow that as offset==1
+          // (i.e. the type)
+          JIT_ASSERT(offset == 1);
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (commutes_with_SumToSize.find(user)) {
+          // these are expressions that might occur during autotdiff operating
+          // on the gradient (matmul would likely be, too but we don't fuse it)
+          // note that for mul, we know (from the chain rule) that only one
+          // factor will be stemming from a calculation involving gradients so
+          // we know that we can move SumToSize across it
+          uses_to_process.insert(
+              uses_to_process.end(),
+              user->output()->uses().begin(),
+              user->output()->uses().end());
+        } else if (user->kind() == aten::_grad_sum_to_size) {
+          // this case should not happen due to how we organize the search
+          // and remove nodes
+          JIT_ASSERT(false);
+        } else if (user->kind() == prim::Return) {
+          // only if we don't already have a _grad_sum_to_size for this,
+          // because we want later (in the graph) nodes to be preferrable
+          // to earlier ones and we iterate backwards
+          // The implicit assumption is that if we have several _grad_sumtosizes
+          // "in parallel" (from addition) they are the same.
+          if (outputGradSumToSizes[offset] == -1) {
+            // note: we make the assumption that the sizes are inputs to the
+            // fusion group (rather than something calculated).
+            outputGradSumToSizes[offset] = node->inputs()[1]->offset();
+          }
+        } else if (user->kind() == prim::FusedConcat) {
+          throw std::runtime_error("don't know how to do fused concat yet");
+        } else {
+          throw std::runtime_error(
+              "unknown node type for fusion SumToSize processing");
+        }
```

Comment:
I feel like this is a bit too strong of a condition. `commutes_with_SumToSize` doesn't exhaust all fusable nodes, so in fact you _can_ trigger this case, and what then? We can't land a patch that has no fallback for such a case.

Reply:
That's a pretty strong assumption to be honest. We should still guard against it to make it more future proof. I wouldn't be surprised if we wanted to emit sth like `at::_sigmoid_backward` to use the optimized implementation if it doesn't get fused.

Reply:
Now we fuse only if we can deal with the processing needed, so I think that's cool now.